wmfdata/spark.py

```diff
@@ -99,28 +134,77 @@ def get_session(type="regular", app_name="wmfdata", extra_settings={}):
         )
     )
 
-    if type == "regular":
-        config = REGULAR_SPARK_SETTINGS
-    elif type == "large":
-        config = LARGE_SPARK_SETTINGS
-    # Add in any extra settings, overwriting if applicable
-    config.update(extra_settings)
-    for k, v in config.items():
+    # Apply any provided spark configs.
+    for k, v in spark_config.items():
         builder.config(k, v)
 
     session = builder.getOrCreate()
 
-    # Cancel any existing timeout on the same session, in case an existing
-    # session is being returned
-    cancel_session_timeout(session)
+    if master == 'yarn' and yarn_session_timeout is not None:
```

Comment:
The general way you handle timers seems problematic. Because the timer starts when the session is (re)initialized, its period will overlap with the use of the session and could easily end up killing the session while something is running. In my experience, Spark jobs can definitely run longer than an hour.

Moreover, the timeout essentially makes it impossible to run anything that takes longer than an hour, because you can't manually reset the timer while the long-running thing is blocking the kernel! Theoretically, the minority of users who will use `get_custom_session` can adjust the timeout as necessary, but it will be hard to discover the need _and_ hard to do, since it's not clear in advance how long something will take.

Previously, running `get_session` would cancel any timer but _not_ start one, since it had no control over how the session was being used and could not tell when it was inactive (the docstring advised the user to start timeouts manually when appropriate, but of course it's likely that most people won't follow that advice). 

On the other hand, `run` starts a timer when it _finishes_, because it's the user of the session and knows when it becomes inactive.

I'm not immediately sure how best to handle this. Some options:
- Set the default timeout to something much longer, like a day. This would be less likely to kill something in progress, but it could still happen. For example, if a user opens a session, uses it for a while, leaves, and comes back 23 hours later to keep using it, they're back to having an effective timeout of an hour.
- Give up altogether on timers that start when a session is _initialized_. This will lead to more YARN containers staying open for long periods (although this will not affect people who just use the basic `run` interface) but I don't think the effect is any worse than a Jupyter kernel doing the same thing.
- The Spark cluster itself automatically stops the scheduler process if it is inactive for long enough. I'm not sure if it's possible, but it seems like the ideal solution since that way the timer is always based on inactive time rather than time since creation. We'd need to tweak this code so people understand what has happened when they try to use a session that's been stopped, and it could still lead to data loss in some circumstances, but it is much more reasonable to say that a Spark session _at rest_ is not guaranteed to be permanent.

Reply:
Ah ok, I think I missed some of that.  Alright, let's do your 2nd solution there.  These get session methods will have a timeout param, but it will be disabled by default.  Run will set a timeout.

Reply:
Actually, for simplicity, I made it so run is the only thing that starts a timeout.  The others don't bother.